powerstore/resource_smb_share.go

```diff
@@ -0,0 +1,658 @@
+								"Read",
+								"Full",
+								"Change",
+							}...),
+						},
+					},
+
+					"access_type": schema.StringAttribute{
+						MarkdownDescription: "The access type.",
+						Description:         "The access type.",
+						Required:            true,
+						Validators: []validator.String{
+							stringvalidator.OneOf([]string{
+								"Allow",
+								"Deny",
+							}...),
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+// Configure - defines configuration for smb share resource
+func (r *resourceSMBShare) Configure(ctx context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
+
+	client, ok := req.ProviderData.(*client.Client)
+
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf("Expected *http.Client, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+
+		return
+	}
+
+	r.client = client
+}
+
+// Create - method to create smb share resource
+func (r *resourceSMBShare) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+
+	var plan models.SMBShare
+
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	// Create new smb share
+	smbCreate := r.planToSMBShareCreateParam(plan)
+	var aclParams *gopowerstore.ModifySMBShareACL
+
+	smbCreateResponse, err := r.client.PStoreClient.CreateSMBShare(ctx, smbCreate)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error creating smb share",
+			"Could not create smb share, unexpected error: "+err.Error(),
+		)
+		return
+	}
+	ID := smbCreateResponse.ID
+	// Get SMBShare Details using ID retrieved above
+	SMBShareDetails, err := r.client.PStoreClient.GetSMBShare(context.Background(), ID)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error reading smb share details",
+			"Could not get smb share, unexpected error: "+err.Error(),
+		)
+		return
+	}
+
+	// update state with current details
+	state, diags := r.SMBShareStateACL(SMBShareDetails, gopowerstore.SMBShareACL{})
+	if diags.HasError() {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
+
+	// update state
+	diags = resp.State.Set(ctx, state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Check if ACL needs to be created
+	if !plan.SMBShareACL.IsUnknown() {
+		aclParams, diags = r.planToSMBShareACLUpdate(ctx, plan)
+		if diags.HasError() {
+			resp.Diagnostics.Append(diags...)
+			return
+		}
+		// Update SMBShare ACL
+		diags = r.updateSMBShareACL(ctx, resp, ID, aclParams)
+		if diags.HasError() {
+			resp.Diagnostics.Append(diags...)
+			return
+		}
+	}
+	smbShareState, diags := r.getSMBShareDetails(ctx, ID)
+	if diags.HasError() {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
+
+	diags = resp.State.Set(ctx, smbShareState)
```

Comment:
state is getting set twice in the create call?

Reply:
In that case, in the second API failure, add the error to the diagnostics (don't return) and set the state at once

Reply:
updated the code to update state only once